app/drivers/zephyr/battery_voltage_divider.c

```diff
@@ -0,0 +1,214 @@
+/*
+ * Copyright (c) 2020 The ZMK Contributors
+ *
+ * SPDX-License-Identifier: MIT
+ */
+
+#define DT_DRV_COMPAT zmk_battery_voltage_divider
+
+#include <device.h>
+#include <drivers/kscan.h>
+#include <drivers/gpio.h>
+#include <drivers/adc.h>
+#include <drivers/sensor.h>
+#include <logging/log.h>
+#include <math.h>
+
+LOG_MODULE_DECLARE(zmk, CONFIG_ZMK_LOG_LEVEL);
+
+#if DT_HAS_COMPAT_STATUS_OKAY(DT_DRV_COMPAT)
+
+struct io_channel_config {
+    const char *label;
+    uint8_t channel;
+};
+
+struct gpio_channel_config {
+    const char *label;
+    uint8_t pin;
+    uint8_t flags;
+};
+
+struct bvd_config {
+    struct io_channel_config io_channel;
+    struct gpio_channel_config power_gpios;
+    uint32_t output_ohm;
+    uint32_t full_ohm;
+};
+
+struct bvd_data {
+    struct device *adc;
+    struct device *gpio;
+    struct adc_channel_cfg acc;
+    struct adc_sequence as;
+    uint16_t adc_raw;
+    uint16_t voltage;
+    uint8_t state_of_charge;
+};
+
+static uint8_t lithium_ion_mv_to_pct(int16_t bat_mv) {
+    // Magic function that maps mV to this discharge graph from adafruit:
+    // https://learn.adafruit.com/li-ion-and-lipoly-batteries/voltages
+    return round(106.818 +
```

Comment:
I am wondering if it is really worth to have something like this.
Converting a read voltage to a percent is quite complicated as it depends on a lot of factors (the type of battery, the temperature, the humidity, the current drawn, the number of cycles the battery went through, ...).
In order to have something more or less accurate, an actual battery monitor chip would be required.

Introducing a dependency on `CONFIG_NEWLIB_LIBC` to be able to perform some exponential calculation in order to make a mapping to something that will not be accurate seems overkill.

IMHO just reading the voltage (using a device tree defined voltage divider as you did) and considering the discharge to be linear would not be a much worst approximation. For example considering 4.2V to be 100% and 3.4V to be 0%.

Reply:
While I do think that this function is more accurate than a simple linear approximation, I agree that the amount it's more accurate probably isn't worth the newlib dependency. Good point on that. I'll readjust (Probably 4.2 to 3.45 to err on the side of caution for people playing with low %s).

Reply:
Honestly I think that it should be quite easy to have conditions where the linear approx is actually better. But it is true than during the mostly non-linear part of the discharge, i.e the very beginning/very end, the exponential approx is most likely more accurate.
Sounds good to have 3.45 as a low threshold.